runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowExecutionStateSampler.java

```diff
@@ -117,7 +117,8 @@ public Optional<ActiveMessageMetadata> getActiveMessageMetadataForWorkId(String
     return Optional.ofNullable(null);
   }
 
-  public Map<String, IntSummaryStatistics> getProcessingDistributionsForWorkId(String workId) {
+  public synchronized Map<String, IntSummaryStatistics> getProcessingDistributionsForWorkId(
```

Comment:
since removeTracker isn't synchronized it still seems like the race is possible.

An alternative since the activeTrackersByWorkId is a concurrent map would be to just call get() first and replace the containsKey check with comparing if null is returned.

similarly the completedProcessingMetrics.containsKey check could be replaced by using getOrDefault

Reply:
yeah i like that better. done!